### ISSUE TITLE
Subtraction operator adds phantom constraints that are unsatisfiable

### ISSUE THREAD

**Issue by brandonroth** [robb/Cartography#74]
When trying to pin a view to the left, right, bottom, and top in a layout block using the subtraction operator or the right or bottom (so the view is X pixels from the right side of it's superview) will cause a random constraint to be added.  This only appears to manifest itself when it's a view inside another autolayout view.  The constraint that gets added sets the superview to a width or height of zero which makes the constraints unsatisfiable.  

Workaround:  Order off adding the constraints appears to matter.  If view A is put inside view B, then if view B has it's constraints added before A's then the phantom constraint isn't added.  In practice this workaround is pretty clunky.  Adding the same constraints using the NSLayoutConstraint methods doesn't exhibit this behavior.  

I put together a sample project to recreate the issue https://github.com/brandonroth/potential-octo-meme.git

**Comment by robb** [comment]
First of all, thanks for providing a way to recreate that issue :sparkling_heart: 

There are some subtleties to this, let me try to address them one by one:

> Order off adding the constraints appears to matter. If view A is put inside view B, then if view B has it's constraints added before A's then the phantom constraint isn't added. In practice this workaround is pretty clunky.

I've been able to confirm this. if `contentView` does not have a superview of its own when it is being layouted, that extra `NSLayoutConstraint` is added, even if `translatesAutoresizingMaskIntoConstraints` is explicitly set to `false`.

However, I'm not sure that I agree that setting up the view hierarchy before adding constraints is clunky, considering that `NSLayoutConstraints` may leak outside a sub-tree and affect the entire window. I'd probably suggest the opposite, trying to always set up the view hierarchy before adding constraints where possible. Ultimately, it's your decision though.

 > Adding the same constraints using the NSLayoutConstraint methods doesn't exhibit this behavior. 

This is true, but the two code snippets in your example are not 100% equivalent.

This

```swift
layout(label) { label in
    label.left == label.superview!.left

    label.right == label.superview!.right - 10

    label.top == label.superview!.top
    label.bottom == label.superview!.bottom
}
```

will call `layoutIfNeeded` on `label.superview` once the constraints are installed. That call, since `label.superview` does not yet have a superview of its own, will install that phantom constraint as outlined above.

On the other hand, your alternative solution

```swift
self.contentView.addConstraint(NSLayoutConstraint(item: label, attribute: .Left, relatedBy: .Equal, toItem: self.contentView, attribute: .Left, multiplier: 1, constant: 0))
self.contentView.addConstraint(NSLayoutConstraint(item: label, attribute: .Right, relatedBy: .Equal, toItem: self.contentView, attribute: .Right, multiplier: 1, constant: -10))
self.contentView.addConstraint(NSLayoutConstraint(item: label, attribute: .Top, relatedBy: .Equal, toItem: self.contentView, attribute: .Top, multiplier: 1, constant: 0))
self.contentView.addConstraint(NSLayoutConstraint(item: label, attribute: .Bottom, relatedBy: .Equal, toItem: self.contentView, attribute: .Bottom, multiplier: 1, constant: 0))
```

will only set up the constraints. Since that code is almost immediately followed by `self.view.addSubviews([contentView])`, layouting will occur after `contentView` has been added to `view`, not resulting in a phantom constraint.

If you'd like to set up the constraints before adding the subviews, you would need to defer the layout pass. You can do this by using Cartography's `constrain` function like so:

```swift
constrain(label) { label in
    label.left == label.superview!.left

    label.right == label.superview!.right - 10

    label.top == label.superview!.top
    label.bottom == label.superview!.bottom
}
```

That seems to work fine in the example project you provided :ok_hand:

**Comment by brandonroth** [comment]
Using constrain works in my situation and eliminates the problem.  Thanks for cluing me into the constrain method.

 think the phantom constrain thing is either a bug / feature / something else in the layout engine itself because I experienced it again when working with a dynamically sized tableview header.  My problem there is that I had called layoutIfNeeded on my header before adding it to the tableview and I was seeing the same phantom constraint.
